lib/ddtrace/profiling/collectors/stack.rb

```diff
@@ -0,0 +1,111 @@
+require 'ddtrace/profiling/events/stack'
+require 'ddtrace/utils/time'
+require 'ddtrace/worker'
+require 'ddtrace/workers/polling'
+
+module Datadog
+  module Profiling
+    module Collectors
+      # Pulls and exports profiling data on an async interval basis
+      class Stack < Worker
+        include Workers::Polling
+
+        MIN_INTERVAL = 0.01
+        MAX_FRAMES = 128
+
+        attr_reader \
+          :ignore_thread,
+          :last_wall_time,
+          :max_frames,
+          :max_time_usage_pct,
+          :recorder
+
+        def initialize(recorder, options = {})
+          @recorder = recorder
+          @max_frames = options.fetch(:max_frames, MAX_FRAMES)
+          @ignore_thread = options.fetch(:ignore_thread, nil)
+          @max_time_usage_pct = options.fetch(:max_time_usage_pct, 2.0)
+
+          # Workers::Async::Thread settings
+          # Restart in forks by default
+          self.fork_policy = options.fetch(:fork_policy, Workers::Async::Thread::FORK_POLICY_RESTART)
+
+          # Workers::IntervalLoop settings
+          self.loop_base_interval = options.fetch(:interval, MIN_INTERVAL)
+
+          # Workers::Polling settings
+          self.enabled = options.fetch(:enabled, false)
+        end
+
+        def start
+          perform
+        end
+
+        def perform
+          collect_and_wait
+        end
+
+        def loop_back_off?
+          false
+        end
+
+        def collect_and_wait
+          run_time = Datadog::Utils::Time.measure do
+            collect_events
+          end
+
+          # Update wait time to throttle profiling
+          self.loop_wait_time = compute_wait_time(run_time)
+        end
+
+        def collect_events
+          @last_wall_time ||= Datadog::Utils::Time.get_time
+          events = []
+
+          # Compute wall time interval
+          current_wall_time = Datadog::Utils::Time.get_time
+          wall_time_interval_ns = (current_wall_time - last_wall_time) * 1e9
+          @last_wall_time = current_wall_time
+
+          # Collect backtraces from each thread
+          Thread.list.each do |thread|
+            next unless thread.alive?
+            next if ignore_thread.is_a?(Proc) && ignore_thread.call(thread)
+
+            event = collect_thread_event(thread, wall_time_interval_ns)
+            events << event unless event.nil?
+          end
+
+          # Send events to recorder
+          recorder.push(events) unless events.empty?
+
+          events
+        end
+
+        def collect_thread_event(thread, wall_time_interval_ns)
+          locations = thread.backtrace_locations
```

Comment:
We can limit `max_frames` here by calling `backtrace_locations(0, max_frames)`.

More information here: https://ruby-doc.org/core-2.5.0/Kernel.html#method-i-caller_locations (`Thread#backtrace_locations` has the same signature as `Kernel#caller_locations`)

Reply:
Hmm, but then we don't know the original `stack_size`... is it worth knowing the original `stack_size` it at the expense of loading data we don't need from `backtrace_locations`? @jd

Reply:
Yeah, this is exactly why I did it this way @marcotc 

It is worth considering if we're limiting stack frames like this, we're paying the memory price for all of them anyways. I think the primary motivation is that we don't want to send more frames over the wire, not memory, and we need to know how many we omitted.

Another consideration; if we're going to trim out less meaningful frames (like `in block`) then we might have to do this anyways.

Reply:
What @delner said basically.

We want to know the total size of the stack trace so we can show that it has been truncated and how many frames we missed.